From 83d0d5b0f35d60a6510f128c96a63ee8ef91dc61 Mon Sep 17 00:00:00 2001
From: Igal Zeltser <izeltser@habana.ai>
Date: Thu, 13 Jan 2022 18:45:11 +0200
Subject: [PATCH] [SW-69581] gaudi3: create file tree

As a preparation of preboot build for gaudi3, minimal required source
files are added for asic onix and pldm.
The source files are empty and will be filled with content later.
In addition, minimal makefiles and linker files are added as well.

At the moment preboot build is panned for architecture Arc only.

Change-Id: I864006e27717742b5be595fef71e286d8dcfde5b
Signed-off-by: Igal Zeltser <izeltser@habana.ai>
---

diff --git a/arch/gaudi3/arc/Makefile b/arch/gaudi3/arc/Makefile
new file mode 100644
index 0000000..2cfbb29
--- /dev/null
+++ b/arch/gaudi3/arc/Makefile
@@ -0,0 +1,6 @@
+FW_SIZE := 0x8000
+PPBOOT_MAX_SIZE := 16384
+PREBOOT_MAX_SIZE := 130048
+
+image_validation: image_validation_common
+image_validation: image_validation_preboot
diff --git a/arch/gaudi3/arc/asic.c b/arch/gaudi3/arc/asic.c
new file mode 100644
index 0000000..14add40
--- /dev/null
+++ b/arch/gaudi3/arc/asic.c
@@ -0,0 +1,4 @@
+// SPDX-License-Identifier:     GPL-2.0+
+/*
+ * (C) Copyright 2022 HabanaLabs, Ltd.
+ */
diff --git a/arch/gaudi3/arc/common.c b/arch/gaudi3/arc/common.c
new file mode 100644
index 0000000..6927945
--- /dev/null
+++ b/arch/gaudi3/arc/common.c
@@ -0,0 +1,6 @@
+// SPDX-License-Identifier:     GPL-2.0+
+/*
+ * (C) Copyright 2022 HabanaLabs, Ltd.
+ */
+
+#include "common.h"
diff --git a/arch/gaudi3/arc/common.h b/arch/gaudi3/arc/common.h
new file mode 100644
index 0000000..14a8c48
--- /dev/null
+++ b/arch/gaudi3/arc/common.h
@@ -0,0 +1,12 @@
+/* SPDX-License-Identifier:     GPL-2.0+
+ *
+ * Copyright (C) 2022 HabanaLabs Ltd.
+ */
+
+#ifndef COMMON_GAUDI3_ARC_GNU_H_
+#define COMMON_GAUDI3_ARC_GNU_H_
+
+#define PPBOOT_SIZE			0x4000		/* 16KB */
+#define PREBOOT_SIZE			0x20000		/* 128KB */
+
+#endif /* COMMON_GAUDI3_ARC_GNU_H_ */
diff --git a/arch/gaudi3/arc/onix.c b/arch/gaudi3/arc/onix.c
new file mode 100644
index 0000000..14add40
--- /dev/null
+++ b/arch/gaudi3/arc/onix.c
@@ -0,0 +1,4 @@
+// SPDX-License-Identifier:     GPL-2.0+
+/*
+ * (C) Copyright 2022 HabanaLabs, Ltd.
+ */
diff --git a/arch/gaudi3/arc/pldm.c b/arch/gaudi3/arc/pldm.c
new file mode 100644
index 0000000..14add40
--- /dev/null
+++ b/arch/gaudi3/arc/pldm.c
@@ -0,0 +1,4 @@
+// SPDX-License-Identifier:     GPL-2.0+
+/*
+ * (C) Copyright 2022 HabanaLabs, Ltd.
+ */
diff --git a/arch/gaudi3/arc/ppboot.lds.S b/arch/gaudi3/arc/ppboot.lds.S
new file mode 100644
index 0000000..93d8b69
--- /dev/null
+++ b/arch/gaudi3/arc/ppboot.lds.S
@@ -0,0 +1,78 @@
+/* SPDX-License-Identifier: GPL-2.0+ */
+/*
+ * Copyright 2022 HabanaLabs, Ltd.
+ * All Rights Reserved.
+ */
+
+#include "common.h"
+
+#ifdef __arc__
+OUTPUT_FORMAT("elf32-littlearc", "elf32-bigarc", "elf32-littlearc")
+OUTPUT_ARCH(arc)
+
+MEMORY
+{
+	.sram 	: ORIGIN = PPBOOT_TEXT_BASE, LENGTH = PPBOOT_SIZE
+	DCCM	: ORIGIN = PPBOOT_DCCM_BASE, LENGTH = 8K
+}
+
+REGION_ALIAS("REGION_DATA",	DCCM)
+REGION_ALIAS("REGION_SDATA",	DCCM)
+REGION_ALIAS("REGION_BSS",	DCCM)
+REGION_ALIAS("REGION_STACK",	DCCM)
+
+#endif
+
+ENTRY(_start)
+SECTIONS
+{
+	.text : {
+		. = ALIGN(8);
+		app/ppboot/crt0_ppboot.o (.text*)
+		*(.text .text.*)
+		_etext = . ;
+	} >.sram
+
+	.rodata : AT ( LOADADDR(.text) + SIZEOF(.text) ) {
+		_rodata = .;
+		*(.rodata .rodata.*)
+		_erodata = . ;
+        } >.sram
+
+	.data : AT ( LOADADDR(.rodata) + SIZEOF(.rodata) ) {
+		. = ALIGN(8);
+		_data = .;
+		*(.data .data.*)
+		_edata = .;
+        } > REGION_DATA
+
+	.sdata : AT ( LOADADDR(.data) + SIZEOF(.data) ) {
+		. = ALIGN(8);
+		_sdata = .;
+		*(.sdata .sdata.* .gnu.linkonce.s.*)
+		_esdata = . ;
+        } > REGION_SDATA
+
+	.bss : {
+		. = ALIGN(8);
+		__bss_start = . ;
+		*(.bss*)
+		__bss_end = . ;
+	} > REGION_BSS
+
+	.stack : {
+		. = ALIGN(8);
+		__stack = . ;
+		. = . + 512 ;
+		__stack_top = . ;
+	} > REGION_STACK
+
+	.stack2 : {
+		. = ALIGN(8);
+		__stack2 = . ;
+		. = . + 512 ;
+		__stack2_top = . ;
+	} > REGION_STACK
+
+	.note.gnu.build-id : { *(.note.gnu.build-id) } > .sram
+}
diff --git a/arch/gaudi3/arc/preboot.lds.S b/arch/gaudi3/arc/preboot.lds.S
new file mode 100644
index 0000000..1d84c39
--- /dev/null
+++ b/arch/gaudi3/arc/preboot.lds.S
@@ -0,0 +1,39 @@
+/* SPDX-License-Identifier:     GPL-2.0+ */
+/*
+ * (C) Copyright 2022 HabanaLabs, Ltd.
+ *
+ */
+
+#include "common.h"
+
+#ifdef __arc__
+MEMORY { .sram : ORIGIN = PREBOOT_TEXT_BASE, LENGTH = PREBOOT_SIZE }
+OUTPUT_FORMAT("elf32-littlearc", "elf32-bigarc", "elf32-littlearc")
+OUTPUT_ARCH(arc)
+#endif
+
+ENTRY(_start)
+SECTIONS
+{
+	.text : {
+		. = ALIGN(8);
+		app/preboot/crt0_preboot.o (.text*)
+		* (.text*)
+	} >.sram
+
+	.bss : {
+		. = ALIGN(8);
+		__bss_start = . ;
+		*(.bss*)
+		__bss_end = . ;
+	} > .sram
+
+	.stack : {
+		. = ALIGN(8);
+		__stack = . ;
+		. = . + 4096 ;
+		__stack_top = . ;
+	} > .sram
+
+	.note.gnu.build-id : { *(.note.gnu.build-id) } > .sram
+}
diff --git a/arch/gaudi3/arc/recovery.lds.S b/arch/gaudi3/arc/recovery.lds.S
new file mode 100644
index 0000000..1975024
--- /dev/null
+++ b/arch/gaudi3/arc/recovery.lds.S
@@ -0,0 +1,39 @@
+/* SPDX-License-Identifier: GPL-2.0+ */
+/*
+ * Copyright 2022 HabanaLabs, Ltd.
+ * All Rights Reserved.
+ */
+
+#include "common.h"
+
+#ifdef __arc__
+MEMORY { .sram : ORIGIN = RECOVERY_TEXT_BASE, LENGTH = RECOVERY_SIZE }
+OUTPUT_FORMAT("elf32-littlearc", "elf32-bigarc", "elf32-littlearc")
+OUTPUT_ARCH(arc)
+#endif
+
+ENTRY(_start)
+SECTIONS
+{
+	.text : {
+		. = ALIGN(4);
+		crt0_recovery.o (.text*)
+		* (.text*)
+	} >.sram
+
+	.bss : {
+		. = ALIGN(4);
+		__bss_start = . ;
+		*(.bss*)
+		__bss_end = . ;
+	} > .sram
+
+	.stack : {
+		. = ALIGN(4);
+		__stack = . ;
+		. = . + 512 ;
+		__stack_top = . ;
+	} > .sram
+
+	.note.gnu.build-id : { *(.note.gnu.build-id) } > .sram
+}
diff --git a/arch/gaudi3/arc/secondary.S b/arch/gaudi3/arc/secondary.S
new file mode 100644
index 0000000..8c15cd3
--- /dev/null
+++ b/arch/gaudi3/arc/secondary.S
@@ -0,0 +1,18 @@
+/* SPDX-License-Identifier:     GPL-2.0+ */
+/*
+ * Copyright (C) 2022 HabanaLabs Ltd.
+ * All Rights Reserved.
+ *
+ */
+
+.global secondary_entry
+secondary_entry:
+	/* stack top set to secondary CPU DCCM 1k offset (internal address) */
+	mov	sp, __stack2_top
+	mov	fp, sp
+
+	/* Disable un-aligned checks on data memory */
+	lr	r0, [STATUS32]
+	bset	r0, r0, 0x13
+	kflag	r0
+	bl	secondary_main
diff --git a/arch/gaudi3/common/asic.c b/arch/gaudi3/common/asic.c
new file mode 100644
index 0000000..64e238b
--- /dev/null
+++ b/arch/gaudi3/common/asic.c
@@ -0,0 +1,4 @@
+// SPDX-License-Identifier:     GPL-2.0+
+/*
+ * Copyright (C) 2022 HabanaLabs Ltd.
+ */
diff --git a/arch/gaudi3/common/common.c b/arch/gaudi3/common/common.c
new file mode 100644
index 0000000..64e238b
--- /dev/null
+++ b/arch/gaudi3/common/common.c
@@ -0,0 +1,4 @@
+// SPDX-License-Identifier:     GPL-2.0+
+/*
+ * Copyright (C) 2022 HabanaLabs Ltd.
+ */
diff --git a/arch/gaudi3/common/onix.c b/arch/gaudi3/common/onix.c
new file mode 100644
index 0000000..64e238b
--- /dev/null
+++ b/arch/gaudi3/common/onix.c
@@ -0,0 +1,4 @@
+// SPDX-License-Identifier:     GPL-2.0+
+/*
+ * Copyright (C) 2022 HabanaLabs Ltd.
+ */
diff --git a/arch/gaudi3/common/pldm.c b/arch/gaudi3/common/pldm.c
new file mode 100644
index 0000000..64e238b
--- /dev/null
+++ b/arch/gaudi3/common/pldm.c
@@ -0,0 +1,4 @@
+// SPDX-License-Identifier:     GPL-2.0+
+/*
+ * Copyright (C) 2022 HabanaLabs Ltd.
+ */
